# Importing the needed packages:

In [1]:
import pandas as pd
import numpy as np

# Data Access and Profiling:
###     Importing the Passengers_Satisfaction_Survey (1) dataset.

In [2]:
df = pd.read_csv('Passengers_Satisfaction_Survey (1).csv')
df

Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
1         Male  Disloyal Customer  20.0  Business travel       Eco   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0               -5421.0                    3.0   
1                   0.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
1                                     3.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
1                   2                    10.0                4          2.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
0                             14.0                       0.0  Dissatisfied  
1                              0.0                       0.0  dissatisfied  
2                              NaN                       NaN       neutral  
3                              0.0                       6.0  Dissatisfied  
4                              NaN                       NaN       neutral  
...                            ...                       ...           ...  
104905                         8.0                       0.0     satisfied  
104906                        54.0                      40.0     satisfied  
104907                         0.0                       2.0     satisfied  
104908                        

# Creating a subset of the DataFrame including only Business Class passenger satisfaction data:

In [3]:
df_Business = df[(df["Class"] == "Business")] 
df_Business

Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
5       Female     Loyal Customer  57.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0               -5421.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
5                   0.0                    3.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
5                                     4.0                      4   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
5                   5                     2.0                3          2.0   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
0                             14.0                       0.0  Dissatisfied  
2                              NaN                       NaN       neutral  
3                              0.0                       6.0  Dissatisfied  
4                              NaN                       NaN       neutral  
5                              0.0                       0.0  Dissatisfied  
...                            ...                       ...           ...  
104905                         8.0                       0.0     satisfied  
104906                        54.0                      40.0     satisfied  
104907                         0.0                       2.0     satisfied  
104908                        

### This is the subsetted, new dataframe that selects only the business class passengers.

# Number of Observations of the data:

In [4]:
df_Business.shape

(50000, 16)

### We have 50,000 rows and 16 columns in this dataframe.

# The name and data type of each data column:

In [5]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  ob

### We have 8 columns with float data types and 8 columns with object data types, with null values in all columns except the Class column.

In [6]:
columns = df_Business.columns
columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Food and drink', 'Inflight entertainment', 'Baggage handling',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'Satisfaction'],
      dtype='object')

### Above are all the columns in the dataframe.

# The unique values of each column:

In [7]:
for x in columns:
    print(x)

    unique = df_Business[x].unique()
    print(unique, '\n*************************************\n\n')


Gender
['Male' 'Female' 'F' 'M' nan] 
*************************************


Customer Type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*************************************


Age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
*************************************


Type of Travel
['Personal Travel' 'Business travel' 'Business' nan] 
*************************************


Class
['Business'] 
*************************************


Flight Distance
[-5.421e+03  0.000e+00  5.600e+01 ...  4.983e+03  2.820e+06  9.000e+06] 
*************************************


Inflight wifi service
[ 3.  1.  2.  4.  0.  5. nan] 
***************************

### Here, we can see all the unique values in each column, which we will clean later on.

# The number of missing values of each column:

In [8]:
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

### As previously stated, we have a lot of missing/null values in each of the columns except the 'class' column.

# Summary statistics for each data column:

In [9]:
df_Business.describe(include='all')

Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       4               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23001           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934737e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes  Arrival Delay in Minutes Satisfaction  
count                 43649.000000              42098.000000        47840  
unique                         NaN                       NaN            4  
top                            NaN                       NaN    satisfied  
freq                           NaN                       NaN        30400  
mean                     13.595363                 13.059647          NaN  
std                      36.595048                 38.116530          NaN  
min                      -2.000000                -45.000000          NaN  
25%                       0.000000                  0.000000          NaN  
50%                       0.000000          

### Above, we can see a summary of all the dataframe. We can see how many unique values each column has, the most repeated value in each column, and also for the "integer" values, we can see the mean, standard deviation and more for each column.

# Frequency distribution (list the number of times each unique value appears) for each data column:

In [10]:
for x in columns:
    print(x)
    value_counts = df_Business[x].value_counts()
    print(value_counts, '\n\n=======================================================')

Gender
Female    23001
Male      22830
F          1361
M          1047
Name: Gender, dtype: int64 

Customer Type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: Customer Type, dtype: int64 

Age
39.0     1935
40.0     1677
41.0     1533
42.0     1514
44.0     1506
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: Age, Length: 74, dtype: int64 

Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
Name: Type of Travel, dtype: int64 

Class
Business    50000
Name: Class, dtype: int64 

Flight Distance
337.0        225
2475.0       187
2586.0       148
1744.0       133
404.0        133
            ... 
1285.0         1
1287.0         1
1288.0         1
1289.0         1
9000000.0      1
Name: Flight Distance, Length: 3573, dtype: int64 

Inflight wifi service
2.0    10784
3.0    10593
4.0     9140
1.0     8808
5.0     7025
0.0     1889
Name: Inflig

### Above, we can see how many times each value is available in each column. This helps us to easily identify strange values such as one value of age being 185 years in the "Age" column, which doesn't make sense.

# The number of fully duplicated data rows:

In [11]:
duplicate = df_Business[df_Business.duplicated()]
  
print("Duplicate Rows :")
duplicate

Duplicate Rows :


Gender      Customer Type   Age   Type of Travel     Class  \
153          F  Disloyal Customer  25.0         Business  Business   
349        NaN                NaN   NaN              NaN  Business   
369        NaN                NaN   NaN              NaN  Business   
705        NaN                NaN   NaN              NaN  Business   
1159    Female     Loyal Customer  49.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104737  Female     Loyal Customer  37.0  Business travel  Business   
104768     NaN                NaN   NaN              NaN  Business   
104789    Male     Loyal Customer  24.0         Business  Business   
104825       M     Loyal Customer  30.0  Business travel  Business   
104840       M     Loyal Customer  30.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
153                67.0                    1.0   
349                 NaN                    NaN   
369                 NaN                    NaN   
705                 NaN                    NaN   
1159              101.0                    1.0   
...                 ...                    ...   
104737           3987.0                    3.0   
104768              NaN                    NaN   
104789           3993.0                    2.0   
104825              NaN                    3.0   
104840              NaN                    4.0   

        Departure/Arrival time convenient Ease of Online booking  \
153                                   0.0                      1   
349                                   NaN                    NaN   
369                                   NaN                    NaN   
705                                   NaN                    NaN   
1159                                  4.0                      4   
...                                   ...                    ...   
104737                                1.0                      3   
104768                                NaN                    NaN   
104789                                5.0                      5   
104825                                3.0                    NaN   
104840                                4.0                    NaN   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
153                 3                     3.0                4          3.0   
349               NaN                     NaN              NaN          NaN   
369               NaN                     NaN              NaN          NaN   
705               NaN                     NaN              NaN          NaN   
1159                3                     2.0                1          3.0   
...               ...                     ...              ...          ...   
104737              5                     5.0                5          5.0   
104768            NaN                     NaN              NaN          NaN   
104789              2                     2.0                A          2.0   
104825            NaN                     4.0                1          4.0   
104840            NaN                     4.0                4          4.0   

        Departure Delay in Minutes  Arrival Delay in Minutes  Satisfaction  
153                            0.0                       0.0           NaN  
349                            NaN                       NaN  dissatisfied  
369                            NaN                       NaN  dissatisfied  
705                            NaN                       NaN  dissatisfied  
1159                          11.0                      14.0  Dissatisfied  
...                            ...                       ...           ...  
104737                        10.0                      23.0     satisfied  
104768                         NaN                       NaN  dissatisfied  
104789                        23.0                     -45.0  dissatisfied  
104825                        

### We have 3,064 duplicated rows which need to be deleted as part of cleaning the dataset.

## Issues with the new dataframe:
1. We had the original "Passengers_Satisfaction_Survey (1)" dataframe with 104,910 rows but we only needed to work on the business class passengers and so we subsetted it to a new dataframe with 50,000 rows.
2. Given that all passengers in the new dataframe are business class, we can consider deleting the "Class" column, it doesn't add any value.
3. Datatypes should be changed (we have many float types that should be integers since they're only numbers(values))
4. We have many unique values, but some of them resemble the same thing (such as 'Male' and 'M' that both resemble the male gender)
5. Some of the values and ratings don't make sense and they should be cleaned/deleted, such as a negative flight distance or a rating of 10 where the threshold is 5.
6. We have many duplicated values in this new dataframe, they need to be deleted.
7. We have 3,064 duplicate rows in the new, subsetted dataframe, and they should be deleted.

# ******* Data Cleansing *******

# Dropping the Class column:

In [12]:
df_Business.drop(['Class'], axis=1, inplace=True)

C:\Users\arman\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [13]:
df_Business = df_Business.reset_index(drop=True)
df_Business

Gender      Customer Type   Age   Type of Travel  Flight Distance  \
0        Male     Loyal Customer  69.0  Personal Travel          -5421.0   
1        Male  Disloyal Customer  41.0  Business travel              0.0   
2        Male     Loyal Customer  52.0  Business travel              0.0   
3      Female  Disloyal Customer  21.0  Business travel              0.0   
4      Female     Loyal Customer  57.0  Business travel              0.0   
...       ...                ...   ...              ...              ...   
49995    Male     Loyal Customer  45.0  Business travel           3305.0   
49996  Female     Loyal Customer  35.0  Business travel            337.0   
49997    Male     Loyal Customer  47.0  Business travel           3496.0   
49998    Male     Loyal Customer  42.0  Business travel           3605.0   
49999  Female           disloyal   0.0  Business travel            862.0   

       Inflight wifi service  Departure/Arrival time convenient  \
0                        3.0                                2.0   
1                        1.0                                1.0   
2                        2.0                                4.0   
3                        4.0                                0.0   
4                        3.0                                4.0   
...                      ...                                ...   
49995                    2.0                                2.0   
49996                    5.0                                5.0   
49997                    2.0                                2.0   
49998                    4.0                                4.0   
49999                    5.0                                5.0   

      Ease of Online booking Food and drink  Inflight entertainment  \
0                          3              5                     1.0   
1                        NaN            NaN                     NaN   
2                          4              1                     2.0   
3                        NaN            NaN                     NaN   
4                          4              5                     2.0   
...                      ...            ...                     ...   
49995                      2              3                     4.0   
49996                      5              2                     4.0   
49997                      2              2                     5.0   
49998                      4              3                     4.0   
49999                      5              5                     5.0   

      Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                    1          3.0                        14.0   
1                  NaN          NaN                         NaN   
2                    2          4.0                         0.0   
3                  NaN          NaN                         NaN   
4                    3          2.0                         0.0   
...                ...          ...                         ...   
49995                4          3.0                         8.0   
49996                4          5.0                        54.0   
49997                5          4.0                         0.0   
49998                4          3.0                         0.0   
49999                1          5.0                         9.0   

       Arrival Delay in Minutes  Satisfaction  
0                           0.0  Dissatisfied  
1                           NaN       neutral  
2                           6.0  Dissatisfied  
3                           NaN       neutral  
4                           0.0  Dissatisfied  
...                         ...           ...  
49995                       0.0     satisfied  
49996                      40.0     satisfied  
49997                       2.0     satisfied  
49998                       0.0     satisfied  
49999                       5.0       neutral  

[50000 rows x 15 columns]

### We now dropped the 'Class' column since all the customers were business class, so it doesn't really help us derive anything.

# Renaming “df_Business” columns by making all the columns’ names in lowercase.  If the column name has whitespace we replace it with an underscore:

In [14]:
df_Business.rename(columns = lambda x: x.lower().strip().replace(' ','_'), inplace= True)

df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

### We have changed the column names to all lower space and whenever there is a space, we changed it to an underscore. The strip also removes any space before or after each of the titles.

# Modifying improperly formatted data and handling bad data:

In [15]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             48239 non-null  object 
 1   customer_type                      48239 non-null  object 
 2   age                                47373 non-null  float64
 3   type_of_travel                     48239 non-null  object 
 4   flight_distance                    46604 non-null  float64
 5   inflight_wifi_service              48239 non-null  float64
 6   departure/arrival_time_convenient  48239 non-null  float64
 7   ease_of_online_booking             42705 non-null  object 
 8   food_and_drink                     42705 non-null  object 
 9   inflight_entertainment             43649 non-null  float64
 10  baggage_handling                   43649 non-null  object 
 11  cleanliness                        43649 non-null  flo

In [16]:
columns = df_Business.columns
columns

for x in columns:
    print(x)

    unique = df_Business[x].unique()
    print(unique, '\n*************************************\n\n')

gender
['Male' 'Female' 'F' 'M' nan] 
*************************************


customer_type
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*************************************


age
[ 69.  41.  52.  21.  57.  45.  39.  33.  10.  55.  59.  53.  56.  47.
  54.  25.  51.  60.  27.  43.  20.  19.  63.  50.  nan  38.  46.  85.
  26.  42.  30.  24.  22.  17.  48.  40.  16.  32.  35.  44.  49.  65.
  67.  66.  64.   7.  37.  36.  15.  62.  34.   0.  68.  70.  58.  11.
  74.  13.  80.  75.   9.  61.  71.  12.  73.  72.  18.  79.   8.  76.
  14.  77.  78. 180. 185.] 
*************************************


type_of_travel
['Personal Travel' 'Business travel' 'Business' nan] 
*************************************


flight_distance
[-5.421e+03  0.000e+00  5.600e+01 ...  4.983e+03  2.820e+06  9.000e+06] 
*************************************


inflight_wifi_service
[ 3.  1.  2.  4.  0.  5. nan] 
*************************************


departure/arrival_time_convenient
[ 2.  1.  4. 

### We have some columns that have the same values with different words, we will combine them to one. These are in the 'gender', 'customer_type', 'type_of_travel', and 'satisfaction' columns. There are some some extreme outlier values that make no sense.  We change the categorical and out of place numerical values to proper numerical values on a scale of 5 to fit the rest of the values in the dataframe. 

In [17]:
df_Business['gender'] = df_Business['gender'].replace(['F','M'],['Female','Male'])
df_Business['customer_type'] = df_Business['customer_type'].replace(['Loyal','disloyal'],['Loyal Customer','Disloyal Customer'])
df_Business['type_of_travel'] = df_Business['type_of_travel'].replace(['Business'],['Business travel'])
df_Business['satisfaction'] = df_Business['satisfaction'].replace(['Dissatisfied'],['dissatisfied'])
df_Business.at[0,'flight_distance'] = abs(df_Business.at[0,'flight_distance'])
df_Business[df_Business['age'] > 179] = np.NaN
df_Business[df_Business['age'] == 0] = np.NaN
df_Business[df_Business['flight_distance'] > 100000] = np.NaN
df_Business['food_and_drink'] = df_Business['food_and_drink'].replace(['Great'],['4'])
df_Business['inflight_entertainment'] = df_Business['inflight_entertainment'].replace([10],[5])
df_Business['baggage_handling'] = df_Business['baggage_handling'].replace(['A'],['5'])
df_Business['ease_of_online_booking'] = df_Business['ease_of_online_booking'].replace('Unknown', np.NaN)


df_Business

gender      customer_type   age   type_of_travel  flight_distance  \
0        Male     Loyal Customer  69.0  Personal Travel           5421.0   
1        Male  Disloyal Customer  41.0  Business travel              0.0   
2        Male     Loyal Customer  52.0  Business travel              0.0   
3      Female  Disloyal Customer  21.0  Business travel              0.0   
4      Female     Loyal Customer  57.0  Business travel              0.0   
...       ...                ...   ...              ...              ...   
49995    Male     Loyal Customer  45.0  Business travel           3305.0   
49996  Female     Loyal Customer  35.0  Business travel            337.0   
49997    Male     Loyal Customer  47.0  Business travel           3496.0   
49998    Male     Loyal Customer  42.0  Business travel           3605.0   
49999     NaN                NaN   NaN              NaN              NaN   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                        3.0                                2.0   
1                        1.0                                1.0   
2                        2.0                                4.0   
3                        4.0                                0.0   
4                        3.0                                4.0   
...                      ...                                ...   
49995                    2.0                                2.0   
49996                    5.0                                5.0   
49997                    2.0                                2.0   
49998                    4.0                                4.0   
49999                    NaN                                NaN   

      ease_of_online_booking food_and_drink  inflight_entertainment  \
0                          3              5                     1.0   
1                        NaN            NaN                     NaN   
2                          4              1                     2.0   
3                        NaN            NaN                     NaN   
4                          4              5                     2.0   
...                      ...            ...                     ...   
49995                      2              3                     4.0   
49996                      5              2                     4.0   
49997                      2              2                     5.0   
49998                      4              3                     4.0   
49999                    NaN            NaN                     NaN   

      baggage_handling  cleanliness  departure_delay_in_minutes  \
0                    1          3.0                        14.0   
1                  NaN          NaN                         NaN   
2                    2          4.0                         0.0   
3                  NaN          NaN                         NaN   
4                    3          2.0                         0.0   
...                ...          ...                         ...   
49995                4          3.0                         8.0   
49996                4          5.0                        54.0   
49997                5          4.0                         0.0   
49998                4          3.0                         0.0   
49999              NaN          NaN                         NaN   

       arrival_delay_in_minutes  satisfaction  
0                           0.0  dissatisfied  
1                           NaN       neutral  
2                           6.0  dissatisfied  
3                           NaN       neutral  
4                           0.0  dissatisfied  
...                         ...           ...  
49995                       0.0     satisfied  
49996                      40.0     satisfied  
49997                       2.0     satisfied  
49998                       0.0     satisfied  
49999                       NaN           NaN  

[50000 rows x 15 columns]

In [18]:
for x in columns:
    print(x)

    unique = df_Business[x].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')


gender
#.of unique values:  3 
List of unique values:  ['Male' 'Female' nan] 
*************************************

customer_type
#.of unique values:  3 
List of unique values:  ['Loyal Customer' 'Disloyal Customer' nan] 
*************************************

age
#.of unique values:  72 
List of unique values:  [69. 41. 52. 21. 57. 45. 39. 33. 10. 55. 59. 53. 56. 47. 54. 25. 51. 60.
 27. 43. 20. 19. 63. 50. nan 38. 46. 85. 26. 42. 30. 24. 22. 17. 48. 40.
 16. 32. 35. 44. 49. 65. 67. 66. 64.  7. 37. 36. 15. 62. 34. 68. 70. 58.
 11. 74. 13. 80. 75.  9. 61. 71. 12. 73. 72. 18. 79.  8. 76. 14. 77. 78.] 
*************************************

type_of_travel
#.of unique values:  3 
List of unique values:  ['Personal Travel' 'Business travel' nan] 
*************************************

flight_distance
#.of unique values:  3570 
List of unique values:  [5421.    0.   56. ... 4817. 4963. 4983.] 
*************************************

inflight_wifi_service
#.of unique values:  7 
List of uniq

# Removing the fully duplicated observations:

In [19]:
df_Business.duplicated().sum()
df_Business.drop_duplicates(inplace=True)
df_Business.duplicated().sum()
df_Business

gender      customer_type   age   type_of_travel  flight_distance  \
0        Male     Loyal Customer  69.0  Personal Travel           5421.0   
1        Male  Disloyal Customer  41.0  Business travel              0.0   
2        Male     Loyal Customer  52.0  Business travel              0.0   
3      Female  Disloyal Customer  21.0  Business travel              0.0   
4      Female     Loyal Customer  57.0  Business travel              0.0   
...       ...                ...   ...              ...              ...   
49992  Female     Loyal Customer  37.0  Business travel           4983.0   
49995    Male     Loyal Customer  45.0  Business travel           3305.0   
49996  Female     Loyal Customer  35.0  Business travel            337.0   
49997    Male     Loyal Customer  47.0  Business travel           3496.0   
49998    Male     Loyal Customer  42.0  Business travel           3605.0   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                        3.0                                2.0   
1                        1.0                                1.0   
2                        2.0                                4.0   
3                        4.0                                0.0   
4                        3.0                                4.0   
...                      ...                                ...   
49992                    2.0                                3.0   
49995                    2.0                                2.0   
49996                    5.0                                5.0   
49997                    2.0                                2.0   
49998                    4.0                                4.0   

      ease_of_online_booking food_and_drink  inflight_entertainment  \
0                          3              5                     1.0   
1                        NaN            NaN                     NaN   
2                          4              1                     2.0   
3                        NaN            NaN                     NaN   
4                          4              5                     2.0   
...                      ...            ...                     ...   
49992                      2              4                     3.0   
49995                      2              3                     4.0   
49996                      5              2                     4.0   
49997                      2              2                     5.0   
49998                      4              3                     4.0   

      baggage_handling  cleanliness  departure_delay_in_minutes  \
0                    1          3.0                        14.0   
1                  NaN          NaN                         NaN   
2                    2          4.0                         0.0   
3                  NaN          NaN                         NaN   
4                    3          2.0                         0.0   
...                ...          ...                         ...   
49992                1          4.0                         0.0   
49995                4          3.0                         8.0   
49996                4          5.0                        54.0   
49997                5          4.0                         0.0   
49998                4          3.0                         0.0   

       arrival_delay_in_minutes  satisfaction  
0                           0.0  dissatisfied  
1                           NaN       neutral  
2                           6.0  dissatisfied  
3                           NaN       neutral  
4                           0.0  dissatisfied  
...                         ...           ...  
49992                       2.0     satisfied  
49995                       0.0     satisfied  
49996                      40.0     satisfied  
49997                       2.0     satisfied  
49998                       0.0     satisfied  

[46117 rows x 15 columns]

### Here, we dropped all the duplicated rows and re-ran the dataframe to see that they have been dropped.

In [20]:
df_Business.isnull().sum()

gender                                  3
customer_type                           3
age                                   864
type_of_travel                          3
flight_distance                       833
inflight_wifi_service                   3
departure/arrival_time_convenient       3
ease_of_online_booking               5102
food_and_drink                       4675
inflight_entertainment               4370
baggage_handling                     4370
cleanliness                          4370
departure_delay_in_minutes           4370
arrival_delay_in_minutes             5209
satisfaction                         2132
dtype: int64

## At this point the only remaining part of data cleansing we have is fixing the NA values, and here is the table of the number of NA values in each column and how we will convert it:

|  Column                            | N of NA values   | NA values converted to  
|  :------                           |   -----          |   -----:
| gender                             |     3            |  Mode 
| customer_type                      |     3            |  Mode
| age                                |   864            |  Mean
| type_of_travel                     |     3            |  Mode
| flight_distance                    |   833            |  Mean
| inflight_wifi_service              |     3            |  Mode
| departure/arrival_time_convenient  |     3            |  Mode
| ease_of_online_booking             |  5102            |  Mode
| food_and_drink                     |  4675            |  Mode
| inflight_entertainment             |  4370            |  Mode
| baggage_handling                   |  4370            |  Mode
| cleanliness                        |  4370            |  Mode
| departure_delay_in_minutes         |  4370            |  Mean
| arrival_delay_in_minutes           |  5209            |  Mean
| satisfaction                       |  2132            |  Mode

In [21]:
df_Business['gender'].fillna(df_Business['gender'].mode()[0], inplace=True)
df_Business['customer_type'].fillna(df_Business['customer_type'].mode()[0], inplace=True)
df_Business['age'].fillna(int(df_Business['age'].mean()), inplace=True)
df_Business['type_of_travel'].fillna(df_Business['type_of_travel'].mode()[0], inplace=True)
df_Business['flight_distance'].fillna(df_Business['flight_distance'].mean(), inplace=True)
df_Business['inflight_wifi_service'].fillna(df_Business['inflight_wifi_service'].mode()[0], inplace=True)
df_Business['departure/arrival_time_convenient'].fillna(df_Business['departure/arrival_time_convenient'].mode()[0], inplace=True)
df_Business['ease_of_online_booking'].fillna(df_Business['ease_of_online_booking'].mode()[0], inplace=True)
df_Business['food_and_drink'].fillna(df_Business['food_and_drink'].mode()[0], inplace=True)
df_Business['inflight_entertainment'].fillna(df_Business['inflight_entertainment'].mode()[0], inplace=True)
df_Business['baggage_handling'].fillna(df_Business['baggage_handling'].mode()[0], inplace=True)
df_Business['cleanliness'].fillna(df_Business['cleanliness'].mode()[0], inplace=True)
df_Business['departure_delay_in_minutes'].fillna(df_Business['departure_delay_in_minutes'].mean(), inplace=True)
df_Business['arrival_delay_in_minutes'].fillna(df_Business['arrival_delay_in_minutes'].mean(), inplace=True)
df_Business['satisfaction'].fillna(df_Business['satisfaction'].mode()[0], inplace=True)

In [22]:
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

# Fixing the columns datatypes:

In [23]:
df_Business = df_Business.convert_dtypes()
df_Business['ease_of_online_booking'] = pd.to_numeric(df_Business['ease_of_online_booking'])
df_Business['food_and_drink'] = pd.to_numeric(df_Business['food_and_drink'])
df_Business['baggage_handling'] = pd.to_numeric(df_Business['baggage_handling'])

df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46117 entries, 0 to 49998
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             46117 non-null  string 
 1   customer_type                      46117 non-null  string 
 2   age                                46117 non-null  Int64  
 3   type_of_travel                     46117 non-null  string 
 4   flight_distance                    46117 non-null  Float64
 5   inflight_wifi_service              46117 non-null  Int64  
 6   departure/arrival_time_convenient  46117 non-null  Int64  
 7   ease_of_online_booking             46117 non-null  int64  
 8   food_and_drink                     46117 non-null  int64  
 9   inflight_entertainment             46117 non-null  Int64  
 10  baggage_handling                   46117 non-null  int64  
 11  cleanliness                        46117 non-null  Int

In [24]:
for x in columns:
    print(x)

    unique = df_Business[x].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')


gender
#.of unique values:  2 
List of unique values:  <StringArray>
['Male', 'Female']
Length: 2, dtype: string 
*************************************

customer_type
#.of unique values:  2 
List of unique values:  <StringArray>
['Loyal Customer', 'Disloyal Customer']
Length: 2, dtype: string 
*************************************

age
#.of unique values:  71 
List of unique values:  <IntegerArray>
[69, 41, 52, 21, 57, 45, 39, 33, 10, 55, 59, 53, 56, 47, 54, 25, 51, 60, 27,
 43, 20, 19, 63, 50, 42, 38, 46, 85, 26, 30, 24, 22, 17, 48, 40, 16, 32, 35,
 44, 49, 65, 67, 66, 64,  7, 37, 36, 15, 62, 34, 68, 70, 58, 11, 74, 13, 80,
 75,  9, 61, 71, 12, 73, 72, 18, 79,  8, 76, 14, 77, 78]
Length: 71, dtype: Int64 
*************************************

type_of_travel
#.of unique values:  2 
List of unique values:  <StringArray>
['Personal Travel', 'Business travel']
Length: 2, dtype: string 
*************************************

flight_distance
#.of unique values:  3570 
List of unique values

In [25]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46117 entries, 0 to 49998
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             46117 non-null  string 
 1   customer_type                      46117 non-null  string 
 2   age                                46117 non-null  Int64  
 3   type_of_travel                     46117 non-null  string 
 4   flight_distance                    46117 non-null  Float64
 5   inflight_wifi_service              46117 non-null  Int64  
 6   departure/arrival_time_convenient  46117 non-null  Int64  
 7   ease_of_online_booking             46117 non-null  int64  
 8   food_and_drink                     46117 non-null  int64  
 9   inflight_entertainment             46117 non-null  Int64  
 10  baggage_handling                   46117 non-null  int64  
 11  cleanliness                        46117 non-null  Int

### After all the cleaning, our dataframe has no null/NaN values, no duplicates, and all the missing values have been replaced too, by either mean or mode as we mentioned that we will do. 

In [26]:
df_Business.to_csv('Project_2_Riachi_Khachatryan_Ike_Group#13.csv')

# Comparing the Data Characteristics before and after the data cleansing

## The data after cleansing

### These changes were made

 - Total data points were reduced from 104910 to 46117 rows as we filter the data based on the "Class" = "Business"
 - After cleansing, many of the data types were changed. Changes are shown in the table below
 
|  Column                            | dtype before cleansing   | dtype after cleansing 
|  :------                           |   -----          |   -----:
| gender                             | object           |  string  
| customer_type                      | object           |  string
| age                                | float64          |  Int64  
| type_of_travel                     | object           |  string
| flight_distance                    | float64          |  Float64
| inflight_wifi_service              | float64          |  Int64  
| departure/arrival_time_convenient  | float64          |  Int64  
| ease_of_online_booking             | object           |  Int64  
| food_and_drink                     | object           |  Int64  
| inflight_entertainment             | float64          |  Int64  
| baggage_handling                   | object           |  Int64  
| cleanliness                        | float64          |  Int64
| departure_delay_in_minutes         | float64          |  Float64
| arrival_delay_in_minutes           | float64          |  Float64
| satisfaction                       | object           |  string 

- We dropped the class column, since all the passengers in the new data frame are business class and therefore this column will not give any added value
- Each dataset had its own unique values which we observed. 
- We deleted 3064 duplicated rows from the original dataframe
- We modifyied improperly formatted data and then replaced NA values either more or mean 
- We  merged similar categories into one, like 'F' and 'M' to 'Female' and 'Male'. This was also part of modifying the improperly formatted data.
- Some of the columns had outlier values which we removed

## Columns names and datatype

## Unique Values for each column

## Number of Null values


## The data before cleansing


## Columns names and datatype


## Unique Values for each column

## Number of Null values